src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -819,5 +832,37 @@ private static ILogger GetLogger(HttpContext httpContext)
                 return loggerFactory.CreateLogger(typeof(RequestDelegateFactory));
             }
         }
+
+        private static void EnsureRequestTaskOfNotNull<T>(Task<T> task)
+        {
+            if (task is null)
+            {
+                throw new InvalidOperationException("Task Result should not be null");
+            }
+        }
+
+        private static void EnsureRequestTaskNotNull(Task task)
+        {
+            if (task is null)
+            {
+                throw new InvalidOperationException("Task Result should not be null");
+            }
+        }
+
+        private static void EnsureRequestValueTaskOfNotNull<T>(ValueTask<T> task)
+        {
+            if (task.Result is null)
+            {
+                throw new InvalidOperationException("ValueTask Result should not be null");
+            }
+        }
+
+        private static void EnsureRequestResultNotNull<T>(T result) where T : IResult
```

Comment:
Why is this generic?

Reply:
This is to take into account classes inheriting from **`IResult`** 🙂  like  
**`private static async Task ExecuteTaskResult<T>(Task<T> task, HttpContext httpContext) where T : IResult`**

Reply:
How does this actually break if you change `T` to just `IResult`? `ExecuteTaskResult` is a little special because it's called dynamically using expression trees. This looks like it's only called from `ExecuteResultWriteResponse` where result is already referenced as just a plain old `IResult`. It looks like this method could just be manually inlined into that method without generics.

Reply:
Ok. In this case we can directly use **`private static async Task ExecuteResultWriteResponse(IResult result, HttpContext httpContext)`**